src/mame/nabu/nabupc.cpp

```diff
@@ -0,0 +1,509 @@
+// license:BSD-3-Clause
+// copyright-holders:Brian Johnson
+/************************************************************************
+ *   NABU PC
+ *
+ *   https://nabu.ca/
+ *
+ *   TODO:
+ *     - Original keyboard support
+ *
+ ***********************************************************************/
+
+#include "emu.h"
+#include "nabupc.h"
+
+#include "bus/centronics/epson_ex800.h"
+#include "bus/centronics/epson_lx800.h"
+#include "bus/centronics/epson_lx810l.h"
+#include "bus/centronics/printer.h"
+#include "bus/nabupc/adapter.h"
+#include "bus/nabupc/keyboard/hlekeyboard.h"
+#include "bus/nabupc/keyboard/keyboard.h"
+#include "bus/rs232/null_modem.h"
+#include "bus/rs232/pty.h"
+#include "bus/rs232/rs232.h"
+#include "machine/clock.h"
+
+#include "nabupc.lh"
+
+//**************************************************************************
+//  UTILITY FUNCTIONS
+//**************************************************************************
+
+// Bit Manipulation
+namespace {
+	template<typename T> constexpr T BIT_SHIFT(unsigned n, unsigned v)
+	{
+		return (T)v << n;
+	}
+
+	template<typename T> void BIT_SET(T& w , unsigned n, unsigned v)
+	{
+		T mask = ~BIT_SHIFT<T>(n, 1);
+		w = (w & mask) | BIT_SHIFT<T>(n, v);
+	}
+}
+
+// Priority Encoder
+/** @brief F9318 input lines */
+typedef enum {
+	PRIO_IN_EI = (1<<8),
+	PRIO_IN_I7 = (1<<7),
+	PRIO_IN_I6 = (1<<6),
+	PRIO_IN_I5 = (1<<5),
+	PRIO_IN_I4 = (1<<4),
+	PRIO_IN_I3 = (1<<3),
+	PRIO_IN_I2 = (1<<2),
+	PRIO_IN_I1 = (1<<1),
+	PRIO_IN_I0 = (1<<0),
+	/* masks */
+	PRIO_I7 = PRIO_IN_I7,
+	PRIO_I6_I7 = (PRIO_IN_I6 | PRIO_IN_I7),
+	PRIO_I5_I7 = (PRIO_IN_I5 | PRIO_I6_I7),
+	PRIO_I4_I7 = (PRIO_IN_I4 | PRIO_I5_I7),
+	PRIO_I3_I7 = (PRIO_IN_I3 | PRIO_I4_I7),
+	PRIO_I2_I7 = (PRIO_IN_I2 | PRIO_I3_I7),
+	PRIO_I1_I7 = (PRIO_IN_I1 | PRIO_I2_I7),
+	PRIO_I0_I7 = (PRIO_IN_I0 | PRIO_I1_I7),
+}   f9318_in_t;
+
+/** @brief F9318 output lines */
+typedef enum {
+	PRIO_OUT_Q0 = (1<<0),
+	PRIO_OUT_Q1 = (1<<1),
+	PRIO_OUT_Q2 = (1<<2),
+	PRIO_OUT_EO = (1<<3),
+	PRIO_OUT_GS = (1<<4),
+	/* masks */
+	PRIO_OUT_QZ = (PRIO_OUT_Q0 | PRIO_OUT_Q1 | PRIO_OUT_Q2)
+}   f9318_out_t;
```

Comment:
You don’t need to `typedef` enums in C++, and avoid using names ending in `_t` if they aren’t nested in a class or in a named namespace (they’re reserved in the global namespace by POSIX).  Also, you can use `DECLARE_ENUM_BITWISE_OPERATORS` from osdcomn.h to avoid a bunch of casts.

Reply:
Hmm it seems using DECLARE_ENUM_BITWISE_OPERATORS causes issues with clang compilation. Since I only use some of the possible bit operations clang complains about unused functions, where as gcc does not.